### PR TITLE
[Submission] Firework Assemblers

### PR DESCRIPTION
**Pull request by AndanteDevs**
Adds a custom crafter that adds easier ways to manipulate Firework Rockets, Stars and adds a Firework Pattern item!

### Firework Assembler
Runs the recipes below. Can by crafted by forming this recipe in a dropper...

... and throwing a Crafting Table along with a Flint and Steel on top.

### Recipes
- *Add duration*

- *Remove duration*

- *Create Firework Pattern from Firework Rocket / Star*

- *Create Firework Pattern and remove explosion data from Firework Rocket / Star*

- *Apply Firework Pattern to Firework Rocket / Star*

- *Apply a Firework Star*

### CODE REVIEW

File: gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#@s = @e[type=item,nbt=CRAFTING_TABLE] at @s if entity @e[type=item,nbt=FLINT_AND_STEEL,distance=..1] if block ~ ~-1 ~ dropper{CC RECIPE} align xyz positioned ~.5 ~-1 ~.5
```
Comment (Kroppeb) [inline gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction]: Use more informal worded headers

File: gm4_firework_assemblers/data/firework_assemblers/functions/main.mcfunction [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+#firework assembler crafter destruction
+execute as @e[type=minecraft:armor_stand,tag=gm4_firework_assembler] at @s unless block ~ ~ ~ minecraft:dropper run function firework_assemblers:destroy
+
+#firework assembler maintenance
+execute as @e[tag=gm4_firework_assembler,type=minecraft:armor_stand] run data merge entity @s {Fire:2000}
+
+#recipes
+execute as @e[tag=gm4_firework_assembler,type=minecraft:armor_stand] at @s if block ~ ~ ~ minecraft:dropper[triggered=false]{Items:[{}]} run function firework_assemblers:recipe_validity_check
```
Comment (Kroppeb) [inline gm4_firework_assemblers/data/firework_assemblers/functions/main.mcfunction]: Maybe have a file called `process.mcfunction` which has these 3 commands (line 5,8 and 11)
so you don't call  the selector (`@e[tag=gm4_firework_assembler,type=minecraft:armor_stand]`) 3 times

File: gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+#Run from main
+
+summon armor_stand ~ ~.1 ~ {Small:1,NoGravity:1,Marker:1,Invulnerable:1,Invisible:1,DisabledSlots:2039552,Tags:["gm4_no_edit","gm4_firework_assembler"],Fire:200000,CustomName:'"gm4_firework_assembler"',ArmorItems:[{},{},{},{id:"minecraft:red_concrete_powder",Count:1}]}
+particle minecraft:flash
+playsound minecraft:entity.firework_rocket.blast player @a[distance=..20] ~ ~ ~ 100 1 1
+data merge block ~ ~ ~ {Items:[],CustomName:'"Firework Assembler"'}
+advancement grant @a[distance=..4] only gm4:firework_assemblers
+execute positioned ~ ~1 ~ run kill @e[type=minecraft:item,distance=..1]
```
Comment (Kroppeb) [inline gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction]: This is killing all items?

File: gm4_firework_assemblers/data/firework_assemblers/functions/destroy.mcfunction [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#@s = @e[type=armor_stand,tag=gm4_custom_crafter] at @s unless block ~ ~ ~ air
+#Run from main
+
+loot spawn ~ ~.6 ~ loot firework_assemblers:blocks/firework_assembler
+particle minecraft:block minecraft:tnt ~ ~1 ~ .1 .25 .1 .05 30 normal @a
+particle minecraft:flash
+playsound minecraft:entity.firework_rocket.blast player @a[distance=..20] ~ ~ ~ 100 1 1
+kill @s[type=!minecraft:player]
```
Comment (Kroppeb) [inline gm4_firework_assemblers/data/firework_assemblers/functions/destroy.mcfunction]: Is this from when you were debugging? I don't see why this check is done.

File: gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction (hunk 1 above)
Comment (AndanteDevs) [reply gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction]: I just copied this from CC. I'll actually change all of my headers to the format I usually use, including what you've just mentioned

File: gm4_firework_assemblers/data/firework_assemblers/functions/main.mcfunction (hunk 2 above)
Comment (AndanteDevs) [reply gm4_firework_assemblers/data/firework_assemblers/functions/main.mcfunction]: Again, copied from CC

File: gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction (hunk 3 above)
Comment (AndanteDevs) [reply gm4_firework_assemblers/data/firework_assemblers/functions/create.mcfunction]: I noticed that while testing, just forgot to implement a fix - will do

**Comment by Bluefire-610** [comment]
I assume the pattern can be used to dupe firework patterns? I think players should still have to produce the cores, duping the patterns really defeats the point of having to set up a proper assembly line.

**Comment by AndanteDevs** [comment]
You can't produce the actual colour/explosion data with this crafter (yet)

**Comment by sparks0** [comment]
What is the intended purpose of the remove duration and remove explosion data functions of the crafter? isn't this about assembly not disassembly?

**Comment by AndanteDevs** [comment]
The general purpose of this module is to modify the data of fireworks in general - Firework **Assemblers** was the best name I could come up with

**Comment by sparks0** [comment]
I feel like this module has been submitted with some features missing. ("You can't produce the actual colour/explosion data with this crafter (yet)"). Perhaps a method of crafting a blank firework from raw ingredients too? I think this module will work best when a redstone assembly line can be built around it. Please re-open this PR once additions are made :) (I do think we'll accept this module eventually)

**Comment by Bluefire-610** [comment]
I agree. I think it'd be cool if this would add multiple machines (or one machine that can be set to different modes) to encourage setting up production lines!

**Pull request closed without merging**